astrokat/noisediode.py

```diff
@@ -9,257 +9,419 @@
-    timestamp_off_set = off(kat)
-    user_logger.debug(
-        "DEBUG: off {} ({})".format(timestamp_off_set, time.ctime(timestamp_off_set))
-    )
-    sleeptime = timestamp_off_set - time.time()
-    user_logger.debug("DEBUG: now {}, sleep {}".format(time.time(), sleeptime))
+    user_logger.info('Add lead time of {}s'
+                     .format(lead_time))
+    user_logger.debug('DEBUG: issue command to switch ND on @ {}'
+                      .format(time.time()))
+    user_logger.trace('TRACE: ts before issue nd on command {}'
+                      .format(time.time()))
+    if duration > lead_time:
+        user_logger.trace('TRACE: Trigger duration > lead_time')
+        on_time = _set_time_(lead_time)
+        user_logger.trace('TRACE: now {} ({})'
+                          .format(time.time(),
+                                  time.ctime(time.time())))
+        user_logger.trace('TRACE: on time {} ({})'
+                          .format(on_time,
+                                  time.ctime(on_time)))
+        user_logger.trace('TRACE: delta {}'
+                          .format(on_time - time.time()))
+        # allow lead time for all to switch on simultaneously
+        # timestamp on = now + lead
+        on_time = on(kat, timestamp=on_time)
+        user_logger.debug('DEBUG: on {} ({})'
+                          .format(on_time,
+                                  time.ctime(on_time)))
+        user_logger.debug('DEBUG: fire nd for {}'
+                          .format(duration))
+        sleeptime = min(duration - lead_time, lead_time)
+        user_logger.trace('TRACE: sleep {}'
+                          .format(sleeptime))
+        off_time = on_time + duration
+        user_logger.trace('TRACE: desired off_time {} ({})'
+                          .format(off_time,
+                                  time.ctime(off_time)))
+        user_logger.trace('TRACE: delta {}'
+                          .format(off_time - on_time))
+        user_logger.debug('DEBUG: sleeping for {} [sec]'
+                          .format(sleeptime))
+        time.sleep(sleeptime)
+        user_logger.trace('TRACE: ts after sleep {} ({})'
+                          .format(time.time(),
+                                  time.ctime(time.time())))
+    else:
+        user_logger.trace('TRACE: Trigger duration <= lead_time')
+        cycle_len = _get_max_cycle_len(kat)
+        nd_setup = {'antennas': 'all',
+                    'cycle_len': cycle_len,
+                    'on_frac': float(duration) / cycle_len,
+                    }
+        user_logger.debug('DEBUG: fire nd for {} using pattern'
+                          .format(duration))
+        pattern(kat, nd_setup, lead_time=lead_time)
+        user_logger.debug('DEBUG: pattern set {} ({})'
+                          .format(time.time(),
+                                  time.ctime(time.time())))
+        off_time = _set_time_(lead_time)
```

Comment:
Will the updated implementation not remove the `lead_time = float(lead_time or _DEFAULT_LEAD_TIME)` ambiguity that cause this edge case?

The new implementation is more explicit:
```
def _get_nd_timestamp_(lead_time):
    """Timestamp for ND switch command with lead time
    """
    return time.time() + lead_time
```
replaced the `_set_time_` function

While the explicit `None` evaluation (where needed) does not make any assumptions but test for a definite value

Reply:
Sorry, my comment was clear.  I didn't mean to just comment on that line, but rather the whole else statement, i.e., where `duration <= lead_time`.  Here we use the pattern mode of the digitiser, which means we are limited by the cycle length.

In the (unlikely) case of the user specifying a very long lead time, so that
`lead_time > cycle_len`, the noise diode would be on for longer than we want (in total).  

Consider this scenario:

```python
now = 5  # seconds since the epoch (just for the example)
cycle_len = 20
lead_time = 25
duration = 10
on_time = pattern(kat, nd_setup, lead_time=lead_time)
# on_time will be about now + lead_time = 30
# pattern() only returns after on_time, so current time is say 30.1 sec
# noise diode cycles (10 sec on, 10 sec off):
#     1: on @ 30, off @ 40
#     2: on @ 50, off @ 60
off_time = _get_nd_timestamp_(lead_time)
# off_time will be now + 25, so about 55.1 sec
# so we will already have starting cycle 2 before we turn off the noise
# diode again thus the noise diode would have been on for a combined
# duration of 15.1 seconds, even though duration was specified as 10 seconds.
off_time = off(kat, timestamp=off_time)
```

I think we should at least log a warning in this case that `lead_time > cycle_len`.  May be even better to raise an exception to prevent the user doing it at all.

Reply:
This is a very useful test case thank you @ajoubertza.
Let's see if I can figure this out.

The use case for coping with `lead_time > cycle_len` came from one of the Mario ND test observations
The setup Mario wanted meant that he needed a 5s lead time in order to have all the NDs fire at the same time, but he also wanted them to only contaminate 1 dump. Giving him the ND value for calibration, but also optimising his scan time.
He used high dump rate observations, which at that time was 2s dumps and meant the ND trigger was required to be 2s long for a single dump.

So me focusing on this implementation only solved the problem for small values, and I did not evaluate what happens when the `lead_time` approaches the maximum cycle length limit (20s seemed forever to me).

So given the use case I would like to keep the functionality if it can be achieved within the cycle length. And I am leaving the `else` section in the `trigger` function as is.
What I am going to do is add an evaluation to check `lead_time` against `max_cycle_len` in the `pattern` function.
Even though the issue has been identified through the `trigger` functionality, nothing prevents the user from setting a very long `lead_time` for the `pattern` functionality as well, so I thought it made sense to add it in `pattern`.
```
    if lead_time >= max_cycle_len:
        user_logger.error('Nonstandard ND usage: lead time > max cycle len')
        raise RuntimeError('ND pattern setting cannot be achieved')
```

What it will do now is if the `lead_time >= max_cycle_len`, an error log will be printed in the user progress output, indicating the position in the observation where it happened.
```
Nonstandard ND usage: lead time > max cycle len
```
After which I raise a RuntimeError exception since I do not want to allow any incorrect behaviour that may lead to confusion.
```
RuntimeError: ND pattern setting cannot be achieved
```

Full evaluation output
```
> astrokat-observe.py --yaml test_nd/nd-trigger-short-fail.yaml
2020-05-20 17:24:20Z - Setting up telescope for observation
2020-05-20 17:24:20Z - Observation start up
2020-05-20 17:24:20Z - Running astrokat version - 0.0+unknown.202005201924
2020-05-20 17:24:20Z - Report: noise-diode off at 1589995463.94
2020-05-20 17:24:20Z - Found 1 target(s): 0 from 0 catalogue(s), 0 from default catalogue and 1 as target string(s)
2020-05-20 17:24:20Z - Observation targets are ['azel']
2019-11-14 06:59:59Z - Waiting for observation setup
2019-11-14 07:00:02Z - INIT
2019-11-14 07:00:02Z - Slewing to first target
2019-11-14 07:00:47Z - Slewed to azel at azel (50.3, 43.7) deg
2019-11-14 07:00:47Z - Tracked azel for 0 seconds
2019-11-14 07:00:47Z - Initialising Track target azel for 60.0 sec
2019-11-14 07:00:47Z - Firing noise diode for 10.0s before target observation
2019-11-14 07:00:47Z - Add lead time of 25.0s
2019-11-14 07:00:47Z - Nonstandard ND usage: lead time > max cycle len
2019-11-14 07:00:47Z - Observation clean up
2019-11-14 07:00:47Z - Returning telescope to startup state
2019-11-14 07:00:47Z - Report: noise-diode off at 1573714851.0
Traceback (most recent call last):
  File "/Users/ruby/Projects/gitSDP/astrokat/venv/bin/astrokat-observe.py", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/ruby/Projects/gitSDP/AstroKAT/astrokat/scripts/astrokat-observe.py", line 8, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/Users/ruby/Projects/gitSDP/AstroKAT/astrokat/astrokat/observe_main.py", line 910, in main
    run_observation(opts, kat)
  File "/Users/ruby/Projects/gitSDP/AstroKAT/astrokat/astrokat/observe_main.py", line 735, in run_observation
    targets_visible += observe(session, target, **obs_plan_params)
  File "/Users/ruby/Projects/gitSDP/AstroKAT/astrokat/astrokat/observe_main.py", line 211, in observe
    lead_time=nd_lead)
  File "/Users/ruby/Projects/gitSDP/AstroKAT/astrokat/astrokat/noisediode.py", line 326, in trigger
    on_time = pattern(kat, nd_setup, lead_time=lead_time)
  File "/Users/ruby/Projects/gitSDP/AstroKAT/astrokat/astrokat/noisediode.py", line 389, in pattern
    raise RuntimeError('ND pattern setting cannot be achieved')
RuntimeError: ND pattern setting cannot be achieved
```